src/widget/wknobcomposed.cpp

```diff
@@ -85,6 +116,28 @@ void WKnobComposed::paintEvent(QPaintEvent* e) {
     p.setRenderHint(QPainter::SmoothPixmapTransform);
     p.drawPrimitive(QStyle::PE_Widget, option);
 
+    // We update m_dCurrentAngle since onConnectedControlChanged uses it for
+    // no-op detection.
+    m_dCurrentAngle = m_dMinAngle + (m_dMaxAngle - m_dMinAngle) * getControlParameterDisplay();
+
+    if (m_pRing) {
+        QPainterPath path;
+        int w = width();
+        int h = height();
+        path.moveTo(w/2.0 + m_iMaskXOffset, h/2.0 + m_iMaskYOffset);
```

Comment:
These calculations are hard to understand. Please use intermediate variables with explaining names and add more comments.